Peterburnett/QUTCapstone2020
# Paypal test

Closes #15 

In case someone wants to test:

1. Get your moodle site accessible from the internet.  (with ngrok, do `ngrok http 443` from command line)
2. Find a way to see all incoming requests to your site (with ngrok, go to `http://localhost:4040/inspect/http`)
3. Update the `wwwroot` variable in your moodle site's config file if your site's URL has changed (if using ngrok, ngrok URL can be found from command line or at `http://localhost:4040/status`)
4. Add the line `$CFG->usepaypalsandbox = 1;` in your moodle site's config file
5. Create a paypal developer account and log in
6. Make a business sandbox account and a personal sandbox account (they may already be made for you by default)
7. Create a REST API app (one may already be made for you by default)
8. View the app details, and find its Client ID in the sandbox API credentials
9. Go to the `test_paypal_sdk.php` file, and replace the existing `$sandboxid` value with your own app's client ID (if you don't do this, any payments you test will go to my sandbox store instead, and the IPN will not be sent back properly)
10. Log into the sandbox business account at www.sandbox.paypal.com
11. Go to Account Settings -> Notification -> Instant Payment Notifications and click on Update
12. Edit the notification URL to https://`wwwroot`/admin/tool/paymentplugin/paymentgateway/paypal/ipn.php, replacing `wwwroot` with whatever your Moodle site's URL is
13. Go to https://`wwwroot`/admin/tool/paymentplugin/paymentgateway/paypal/test_paypal_sdk.php?id=`courseid` replacing `wwwroot` with your Moodle site's URL and `courseid` with the ID of an existing course.
14. Click on the paypal button to buy the course, and log in with the sandbox personal account
15. Complete purchase, and check incoming POST requests to the site
16. When a POST request to the IPN page comes, check the values to ensure nothing looks wrong, and check that the response received was a single string "VERIFIED" (IPN page currently performs a var_dump of the response string because the enrolment method is not yet implemented)


Basic requirements for the paypal payments to work in a live environment are:

- User needs a client ID for their app/shop
- User needs to set the IPN page on their paypal business account (the account that receives money)
- User's moodle site needs to be accessible by paypal so that the IPN can be sent to their site


Things that need working on:

- Get the seller's Client ID from subplugin settings once the subplugin settings page is done
- Get the course price from the course settings once the course settings is done
